### PR TITLE
[doctrine-bundle] Docker support

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| License       | MIT

Requires https://github.com/dunglas/symfony-docker-skeleton/pull/5 and https://github.com/symfony/flex/pull/128.

### CODE REVIEW

File: doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml
Comment (Pierstoval) [inline doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml]: Is there any good reason to twitch to pgsql for default projects? Mysql is still the most used for simple projects 😕

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml]: Because Postgres is better,  has more features (JSON queries, advanced pagination and many many more...) and is as easy to use than MySQL.

Comment (OskarStark) [reply doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml]: And has better default values

Comment (maxhelias) [reply doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml]: A PR already exists about this. #411

Comment (fabpot) [reply doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml]: @dunglas  Can you revert this change?

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml]: @fabpot you mean switching back to MySQL?

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (fabpot) [inline doctrine/doctrine-bundle/1.6/manifest.json]: Like for the Messenger recipe, I would let Docker choose a random port for localhost to allow running several projects in parallel.

**Comment by fabpot** [comment]
Anyone willing to finish this one?

**Comment by maxhelias** [comment]
@fabpot I can take care of it if it's OK for everyone

**Comment by dunglas** [comment]
I'm on it @maxhelias

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (maxhelias) [inline doctrine/doctrine-bundle/1.6/manifest.json]: I'm not in favor to put the option of the database in the DSN. cf. https://github.com/symfony/recipes/pull/258#issuecomment-346142034

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (fabpot) [inline doctrine/doctrine-bundle/1.6/manifest.json]: `db` should be reverted to `127.0.0.1`.

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/manifest.json]: If we do that, the recipe will not work "out of the box" anymore

Comment (fabpot) [reply doctrine/doctrine-bundle/1.6/manifest.json]: If we change the name to `db`, the recipe will not work anymore when not using Docker :)

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/manifest.json]: But it’s already not working in this case isn’t it? I mean even with a local installation of MariaDB, the DB name, the login and the password have to be changed.

Comment (bocharsky-bw) [reply doctrine/doctrine-bundle/1.6/manifest.json]: Why not create a commented out env var specifically for Docker like:
```
"#4": "If you use Docker - uncomment the next line"
"#DATABASE_URL": "mysql://symfony:password@db:3306/symfony"
```

And leave the original one unchanged:
```
"DATABASE_URL": "mysql://db_user:db_password@127.0.0.1:3306/db_name"
```

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/manifest.json]: My main goal with the Docker support is to have something totally transparent for the end user, as we’ve done for API Platform.

The user should be able to download the Docker skeleton, run docker-compose up, run docker-compose exec composer require orm, and everything must work seamlessly, without having to tweak config files or even to know how Docker works in the beginning.

@fabpot could it be possible to set a different env var depending if Docker support is enabled or not? By introducing a new key in the manifest for instance?

Comment (maxhelias) [reply doctrine/doctrine-bundle/1.6/manifest.json]: Or we can just set in the environment variable a default value like `{db:-127.0.0.1:3306}`.

After that we will be able in the `EnvConfigurator` to replace if docker support is enabled or not.

```php 
preg_replace('#^{(\w+):-(\w+.+)}$#', $this->composer->getPackage()->getExtra()['symfony']['docker'] ? '${1}' : '${2}', $env);
```

WDYT ?

**Comment by dunglas** [comment]
@fabpot this one is now ready to be merged.

I added a hack in https://github.com/dunglas/symfony-docker/pull/74 to ensure that everything will work just with a simple `composer req doctrine`.

**Review by maxhelias — APPROVED** [review]

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (maxhelias) [reply doctrine/doctrine-bundle/1.6/manifest.json:34]: It's hardcode in Flex : https://github.com/symfony/flex/blob/main/src/Configurator/DockerComposeConfigurator.php#L34-L35

Comment (Pierstoval) [reply doctrine/doctrine-bundle/1.6/manifest.json:34]: Well, it's probably a bit too much, but I'd recommend the creation of a special configurator for a `docker-compose.override.yml.dist` file, so maintainers can explicitly define what can be "publicly" exposed, or what kind of "machine-specific" configuration can be used. 

Exposing public ports is not the best, it can cause conflicts, or priority issues (like two apps listening to the same port), etc., and having them directly in the `docker-compose.yml` file, even commented, might invite users to put the public ports in it.

Having the `dist` file invites the user to create their own use-case when needed, instead of having it enabled by default.

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/manifest.json:34]: > Well, it's probably a bit too much, but I'd recommend the creation of a special configurator for a docker-compose.override.yml.dist file, so maintainers can explicitly define what can be "publicly" exposed, or what kind of "machine-specific" configuration can be used.

@Pierstoval haha, actually I'm already on it. But it's not a blocker for this one. We'll be able to update it to leverage the new capabilities of the configurator later.

> Exposing public ports is not the best, it can cause conflicts, or priority issues (like two apps listening to the same port), etc., and having them directly in the docker-compose.yml file, even commented, might invite users to put the public ports in it.

It will be not be necessary anymore with the new configurator, indeed.

Comment (Pierstoval) [reply doctrine/doctrine-bundle/1.6/manifest.json:34]: > actually I'm already on it

You're always in advance 🤣 👍

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/manifest.json:34]: And here we are: symfony/flex#705

**Review by chalasr — DISMISSED** [review]

**Comment by nicolas-grekas** [comment]
I'd vote +2 if this were using PG instead of mariadb ;)

**Comment by dunglas** [comment]
I'm also totally for using Postgres. It was what I proposed when I opened this PR but it has been refused at this time (3 years ago, see comments above). IIRC @lyrixx was also in favor of using Postgres by default.

Using Postgres instead of MariaDB/MySQL, in addition to the usual benefits, also has the following ones:

* out of the box support for the PG NOTIFY transport for Messenger
* ability for the API Platform distribution, which already ships Postgres, to use this recipe too

WDYT @fabpot?

**Comment by Pierstoval** [comment]
All the docs related to database handling with Symfony will have to change too, if default recipes start using Docker-Compose, especially if MySQL/MariaDB is changed for Postgres.

This might introduce a whole new good practice for newcomers :)

**Comment by fabpot** [comment]
I'm all for psql as well as I'm using it exclusively.
Please, do NOT approve until I test it with Flex thoroughly (thank you).

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (fabpot) [reply doctrine/doctrine-bundle/1.6/manifest.json:34]: I don't understand the issue here. On the localhost, the port is going to be random, so there is no possible conflicts.

Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/manifest.json]: If we put the port in `docker-compose.yml` instead of not in `docker-compose.override.yml` the port will be exposed publicly even in production, which can be a security issue.

**Comment by dunglas** [comment]
I switched to Postgres in the last commit.

**Review by nicolas-grekas — DISMISSED** [review]

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (fabpot) [inline doctrine/doctrine-bundle/1.6/manifest.json:21]: This session is very opinionated.

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (fabpot) [inline doctrine/doctrine-bundle/1.6/manifest.json:25]: should be database

**Review by fabpot — COMMENTED** [review]
@symfony-flex-server review again please

**Comment by fabpot** [comment]
@symfony-flex-server review again please

**Comment by fabpot** [comment]
@symfony-flex-server review again please

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (dunglas) [reply doctrine/doctrine-bundle/1.6/manifest.json:21]: What do you mean? Postgres-opinionated? That's the case, but there are no alternatives as far as I know.

**Review by fabpot — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (pyrech) [inline doctrine/doctrine-bundle/1.6/manifest.json:41]: Shouldn't this service be named `database` too?

Comment (chalasr) [reply doctrine/doctrine-bundle/1.6/manifest.json:41]: correct :) Up for a PR @pyrech?

Comment (pyrech) [reply doctrine/doctrine-bundle/1.6/manifest.json:41]: Yep :slightly_smiling_face:

Comment (pyrech) [reply doctrine/doctrine-bundle/1.6/manifest.json:41]: See https://github.com/symfony/recipes/pull/843 :wink:

Comment (fabpot) [reply doctrine/doctrine-bundle/1.6/manifest.json:41]: You're right, can you do a PR?

File: doctrine/doctrine-bundle/1.6/manifest.json
Comment (alexander-schranz) [inline doctrine/doctrine-bundle/1.6/manifest.json:26]: @dunglas Not sure if I did already ask this 🙈 , but was there a specific reasons why this yaml files are instead of 4 spaces 2 spaces. And so are not similar to symfony config files?
Make it harder in handling .editorconfig files.

Comment (fabpot) [reply doctrine/doctrine-bundle/1.6/manifest.json:26]: 4 spaces are the way to go indeed.

Comment (alexander-schranz) [reply doctrine/doctrine-bundle/1.6/manifest.json:26]: Here we go: https://github.com/symfony/recipes/pull/1018 :)

**Pull request merged**
